Fix Moderated session on leave pause action.

jakule opened the pull request:
Closes https://github.com/gravitational/teleport-private/issues/313

jakule committed:
Fix Moderated session on leave pause action.

jentfoo commented on lib/auth/session_access.go:
```diff
@@ -286,6 +286,17 @@ policySetLoop:
 			continue
 		}
 
+		if options.OnLeaveAction != types.OnSessionLeaveTerminate {
```
This code assumes there are only two states, but the `OnSessionLeaveAction` type is able to accept future additional states.  Although I can't predict what we may in the future this may be a potential bug if we do add another leave action.

We could switch from a string to just a boolean to represent true to one of the possible exit states.

Or switching this to `options.OnLeaveAction == types.OnSessionLeavePause` may be incrementally safer.

jentfoo reviewed (approved):
Changes generally look good, I only noticed one place where I saw any recommendation for improvement

tigrato committed:
Unpause when moderator rejoins session with onLeave=pause

This commite fixes a case when the moderator rejoins a paused session
but Teleport didn't resume the operations and the session was still
locked to everyone.

jakule committed:
Fix panic when re-joining SSH session.

tigrato commented on lib/srv/sess.go:
```diff
@@ -876,7 +880,13 @@ func (s *session) setHasEnhancedRecording(val bool) {
+func (s *session) launch() {
+	// Mark the session as started here, as we want to avoid double initialization.
+	if s.started.Swap(true) {
+		s.log.Debugf("Session has already satrted")
```
```suggestion
		s.log.Debugf("Session has already started")
```

tigrato reviewed (approved):
LGTM

strideynet commented on lib/kube/proxy/sess.go:
```diff
@@ -948,6 +948,18 @@ func (s *session) join(p *party) error {
 			}
 		}
+	} else if canStart && s.tracker.GetState() == types.SessionState_SessionStatePending {
+		// If the session is already running, but the party is a moderator that leaved
```
```suggestion
		// If the session is already running, but the party is a moderator that left
```

strideynet commented on lib/kube/proxy/sess.go:
```diff
@@ -948,6 +948,18 @@ func (s *session) join(p *party) error {
+	} else if canStart && s.tracker.GetState() == types.SessionState_SessionStatePending {
+		// If the session is already running, but the party is a moderator that leaved
+		// a session with onLeave=pause and then rejoined, we need to unpause the session.
+		// When the moderator leaved the session, the session was paused, and we spawn
```
```suggestion
		// When the moderator left the session, the session was paused, and we spawn
```

tigrato committed:
Update lib/kube/proxy/sess.go

Co-authored-by: Noah Stride <noah.stride@goteleport.com>

tigrato committed:
Update lib/kube/proxy/sess.go

Co-authored-by: Noah Stride <noah.stride@goteleport.com>

jakule committed:
Fix typo

Co-authored-by: Tiago Silva <tiago.silva@goteleport.com>
